code/drasil-code/lib/Language/Drasil/Code/Imperative/Modules.hs

```diff
@@ -366,6 +381,15 @@ constrExc c = do
   msgs <- mapM (constraintViolatedMsg q "expected") cs
   return $ map (bodyStatements . (++ [throw "InputError"])) msgs
 
+exc :: (OOProg r) => (CodeVarChunk, ConstraintCE) ->
+  GenState [MSStatement r]
+exc (v, Elem _ e) = do
```

Comment:
I can pattern match on `e` to check that it is a `Set` then otherwise return an error

Reply:
I feel that this shouldn't error-out at all, because whoever calls `exc` should already know it's fine, and thus only pass in the necessary information (so that `exc` doesn't have to pattern match again)

Reply:
I see. Are you saying that `exc` should be directly given the `set` or that it should be able to work if the chunk is constrained to be an `Elem` of something other than a `set`?

Reply:
`exc` should be given the `set`. Whether you want to make it more general is for you to decide if that would be useful.

Reply:
You would need the single caller of `exc` to do that job itself (rather than a separate function). This change should "propagate upwards" through the code.

Reply:
Something like this?
![image](https://github.com/user-attachments/assets/c151d43a-dea0-4044-a4c2-bb44e5cc910f)

Reply:
Something like this. But it might make sense to filter `ch` for `Elem` instead of returning all those `emptyStmt`? In fact, it might make sense to filter *cs* (which is the raw-er input to `chooseConstr`) instead.

Don't think of it as "getting the job done" but rather "putting the code in the most logical place". You're putting the code a little too late in the logic.

Also, you should go back to the top of this particular sub-conversation, which was about digging out the inner list!

Reply:
If I don't add `emptyStmt` I would have to define all of the set variables before the block of if statements. Otherwise, the order in which the set variables are declared will not line up with the order that the constraints appear.

Reply:
With regards to your top comment, why would I need to pattern match on `e`?

Reply:
Re: `emptyStmt` - that makes sense, but the generation of "dummy code" still worries me. Does anything match on `emptyStmt` later on to choose not to emit anything? That would be bad. The information about whether to emit a declaration or not should be directly visible instead of being indirectly encoded.

The pattern match on `e` seems to happen later in the code, when it tries to figure out if things are an explicit list or something else. Again that knowledge should be "more visible" in the data itself.

Reply:
The `emptyStmt` is only used once and I change `exc` to `constrVarDec` which experts a `set` and it doesn't use pattern matching. Would it be better to declare all of the variables beforehand so that having to emit a declaration is unnecessary?